Libraries/Lists/FlatList.js

```diff
@@ -518,8 +520,13 @@ class FlatList<ItemT> extends React.PureComponent<Props<ItemT>, void> {
     }
   };
 
-  _getItemCount = (data: ?Array<ItemT>): number => {
-    if (Array.isArray(data)) {
+  _getItemCount = (data: ?$ArrayLike<ItemT>): number => {
+    if (
+      data &&
+      (typeof data === 'object' || typeof data === 'function') &&
+      typeof data.length === 'number' &&
+      Symbol.iterator in data
```

Comment:
I suggest moving this to `_checkProps`. It seem more like a check for a precondition of the prop than something that needs to be checked potentially multiple times per render: I mean, the call pattern is a discretion of the `VirtualizedList` component and we don't expect `data` to change except when the prop updates.

Reply:
The tricky part of that is for compatibility, we cannot trigger an invariant violation like other checks in `_checkProps`, and instead just want to tell the underlying VirtualizedList not to try to read from the source.

VirtualizedList does assume `getItemCount` is somewhat cheap, so I agree this checks are a little icky to be adding to be run repeatedly. But it avoids introducing state into the list. I think I should be able to make this check marginally cheaper though.